cli_tools/common/distro/distro.go

```diff
@@ -33,155 +36,216 @@ const (
-		return r, errors.New("Windows not yet implemented")
+	hyphen := strings.LastIndex(os, "-")
+	if hyphen < 0 || hyphen == len(os) {
+		return r, fmt.Errorf("expected pattern of `distro-version`. Actual: `%s`", os)
 	}
-	parts := strings.Split(os, "-")
-	distro := parts[0]
-
-	// 1. Format is `(debian|centos|rhel|opensuse)-major`.
-	//   Canonical example: debian-8
-	if distro == debian ||
-		distro == centos ||
-		distro == rhel ||
-		distro == opensuse {
-		if len(parts) != 2 {
-			return r, fmt.Errorf("expected pattern of `distro-version`. Actual: `%s`", os)
+	distro, version := os[:hyphen], os[hyphen+1:]
+
+	if distro == ubuntu {
+		// In gcloud, major and minor are combined as MMmm, such as ubuntu-1804
+		if len(version) != 4 {
+			return r, fmt.Errorf("expected version with length four. Actual: `%s`", version)
 		}
 
-		return release{
-			distro: distro,
-			major:  parts[1],
-		}, nil
+		return FromComponents("ubuntu", version[:2], version[2:])
 	}
+	return FromComponents(distro, version, "")
+}
 
-	// 2. Format is `sles(-variant)?-major`.
-	//   Canonical examples: sles-sap-12, sles-12
-	if distro == sles {
-		switch len(parts) {
-		case 2:
-			return release{
-				distro: distro,
-				major:  parts[1],
-			}, nil
-		case 3:
-			return release{
-				distro:  distro,
-				variant: parts[1],
-				major:   parts[2],
-			}, nil
-		default:
-			return r, fmt.Errorf("unrecognized SLES identifier: `%s`", os)
-		}
+// commonLinuxRelease is a Release that:
+//  1. Has integer major and minor versions.
+//  2. Compatibility is determined by the major version.
+//  3. There are no variants.
+type commonLinuxRelease struct {
+	distro string
+	major  int
+	minor  int
+}
+
+func (r commonLinuxRelease) AsGcloudArg() string {
+	return fmt.Sprintf("%s-%d", r.distro, r.major)
+}
+
+func (r commonLinuxRelease) ImportCompatible(other Release) bool {
+	realOther, ok := other.(commonLinuxRelease)
+	return ok &&
+		r.distro == realOther.distro &&
+		r.major == realOther.major
+}
+
+// The caller is responsible for verifying the syntax of the arguments.
+// Verify the following before calling:
+//   - distro is one of {centos, debian, opensuse, rhel}
+//   - major is >= 1 and minor is >= 0
+func newCommonLinuxRelease(distro string, major, minor int) (Release, error) {
+	verify.GreaterThanOrEqualTo(major, 1)
+	verify.GreaterThanOrEqualTo(minor, 0)
+	verify.Contains(distro, []string{centos, debian, opensuse, rhel})
+	return commonLinuxRelease{
+		distro: distro,
+		major:  major,
+		minor:  minor,
+	}, nil
+}
+
+// slesRelease is a Release that represents the SLES distro and its variants (such as SLES for SAP).
+// Compatibility requires the same variant and major version.
+type slesRelease struct {
+	variant string
+	major   int
+	minor   int
+}
+
+// The caller is responsible for verifying the syntax of the arguments.
+// Verify the following before calling:
+//   - distroAndVariant starts with "sles" or "sles-"
+//   - major is >= 1 and minor is >= 0
```

Comment:
These two types of errors are handled in this func in different ways:
- major/minor via panic
- "sles" prefix via errors

Reply:
The panics are used when a precondition is violated (major/minor, and requiring a prefix of `sles-`, while an error only occurs when the variant name is not recognized. 

The distinction is the burden that we place on the caller; while we expect the caller to give us well-formed version numbers and version strings, we don't expect them to pre-verify the variant name.